### ISSUE TITLE
Siege "BreakableBlocks" are not working in 1.18

### ISSUE THREAD

**Issue by TomLewis** [GriefPrevention/GriefPrevention#1794]
### Observed Behavior

When using /siege, "BreakableBlocks" do not work at all.

### Expected Behavior

To be able to break soft blocks when you siege someone.

### Reproduction steps

1. Enable siege in the worlds you want it
2. /siege a player in their claim
3. Break a soft block, you will be told that the block is too strong or you dont have trust perms.

### Server version

```shell
Current: git-Purpur-1547 (MC: 1.18.1)*\nPrevious: git-Purpur-1541 (MC: 1.18.1)\n* You are running the latest version
```

### GriefPrevention version

```shell
GriefPrevention version 16.18-RC1-4-gfb75955
```

### Configuration

```yaml
# Default values are perfect for most servers.  If you want to customize and have a question, look for the answer here first: http://dev.bukkit.org/bukkit-plugins/grief-prevention/pages/setup-and-configuration/

GriefPrevention:
  SeaLevelOverrides:
    PirateCraft_1: -1
    PirateCraft_1_nether: -1
    PirateCraft_1_the_end: -1
    PirateCraft_1_nether2: -1
    locker: -1
  Claims:
    Mode:
      PirateCraft_1_nether2: Survival
      PirateCraft_1_the_end: Survival
      PirateCraft_1: Survival
      PirateCraft_1_nether: Survival
      locker: Survival
    PreventGlobalMonsterEggs: true
    PreventTheft: false
    ProtectCreatures: true
    PreventButtonsSwitches: true
    LockWoodenDoors: false
    LockTrapDoors: true
    LockFenceGates: true
    EnderPearlsRequireAccessTrust: true
    RaidTriggersRequireBuildTrust: true
    ProtectHorses: true
    ProtectDonkeys: true
    ProtectLlamas: true
    InitialBlocks: 225
    Claim Blocks Accrued Per Hour:
      Default: 60
    Max Accrued Claim Blocks:
      Default: 1000000
    Accrued Idle Threshold: 0
    AccruedIdlePercent: 0
    AbandonReturnRatio: 0.9
    AutomaticNewPlayerClaimsRadius: 7
    AutomaticNewPlayerClaimsRadiusMinimum: 0
    ExtendIntoGroundDistance: 5
    MinimumWidth: 3
    MinimumArea: 80
    MaximumDepth: -2147483648
    InvestigationTool: STICK
    ModificationTool: GOLDEN_SHOVEL
    Expiration:
      ChestClaimDays: 3
      UnusedClaimDays: 14
      AllClaims:
        DaysInactive: 90
        ExceptWhenOwnerHasTotalClaimBlocks: 1000
        ExceptWhenOwnerHasBonusClaimBlocks: 2500
      AutomaticNatureRestoration:
        SurvivalWorlds: false
    AllowTrappedInAdminClaims: true
    MaximumNumberOfClaimsPerPlayer: 0
    CreationRequiresWorldGuardBuildPermission: true
    VillagerTradingRequiresPermission: false
    CommandsRequiringAccessTrust: /sethome;/top;/crew sethome;/etop;/esethome
    DeliverManuals: false
    ManualDeliveryDelaySeconds: 30
    RavagersBreakBlocks: true
    FireSpreadsInClaims: false
    FireDamagesInClaims: false
    LecternReadingRequiresAccessTrust: true
  Spam:
    Enabled: false
    LoginCooldownSeconds: 60
    LoginLogoutNotificationsPerMinute: 5
    ChatSlashCommands: /me;/tell;/global;/local;/w;/msg;/r;/t;/c
    WhisperSlashCommands: /tell;/pm;/r;/w;/whisper;/t;/msg
    WarningMessage: Please reduce your noise level.  Spammers will be banned.
    BanOffenders: true
    BanMessage: Auto-Banned for spam by GriefPrevetion.
    AllowedIpAddresses: http://www.piratemc.com; piratemc.com
    DeathMessageCooldownSeconds: 60
    Logout Message Delay In Seconds: 0
  PvP:
    RulesEnabledInWorld:
      PirateCraft_1: true
      PirateCraft_1_nether: true
      PirateCraft_1_the_end: true
      PirateCraft_1_nether2: true
      locker: true
    ProtectFreshSpawns: false
    PunishLogout: false
    CombatTimeoutSeconds: 14
    AllowCombatItemDrop: true
    BlockedSlashCommands: /home;/vanish;/spawn;/tpa;/top;/etop;/back;/eback;/return;/ereturn;/espawn;/ewarp;/ehome;/kits;/depositall
    ProtectPlayersInLandClaims:
      PlayerOwnedClaims: false
      AdministrativeClaims: false
      AdministrativeSubdivisions: false
    AllowLavaDumpingNearOtherPlayers:
      PvPWorlds: true
      NonPvPWorlds: false
    AllowFlintAndSteelNearOtherPlayers:
      PvPWorlds: true
      NonPvPWorlds: false
    ProtectPetsOutsideLandClaims: false
  Economy:
    ClaimBlocksMaxBonus: 0
    ClaimBlocksPurchaseCost: 1.0
    ClaimBlocksSellValue: 0.5
  ProtectItemsDroppedOnDeath:
    PvPWorlds: false
    NonPvPWorlds: false
  BlockLandClaimExplosions: true
  BlockSurfaceCreeperExplosions: false
  BlockSurfaceOtherExplosions: false
  LimitSkyTrees: true
  LimitTreeGrowth: false
  PistonMovement: CLAIMS_ONLY
  PistonExplosionSound: true
  FireSpreads: true
  FireDestroys: true
  AdminsGetWhispers: false
  AdminsGetSignNotifications: true
  SmartBan: true
  Mute New Players Using Banned Words: true
  MaxPlayersPerIpAddress: 3
  SilenceBans: true
  Siege:
    Worlds:
    - PirateCraft_1
    - PirateCraft_1_nether
    - PirateCraft_1_nether2
    - PirateCraft_1_the_end
    BreakableBlocks:
    - GRASS_BLOCK
    - DIRT
    - COBBLESTONE
    - OAK_PLANKS
    - SPRUCE_PLANKS
    - BIRCH_PLANKS
    - JUNGLE_PLANKS
    - ACACIA_PLANKS
    - DARK_OAK_PLANKS
    - SAND
    - GRAVEL
    - GLASS
    - GRASS
    - FERN
    - DEAD_BUSH
    - WHITE_WOOL
    - ORANGE_WOOL
    - MAGENTA_WOOL
    - LIGHT_BLUE_WOOL
    - YELLOW_WOOL
    - LIME_WOOL
    - PINK_WOOL
    - GRAY_WOOL
    - LIGHT_GRAY_WOOL
    - CYAN_WOOL
    - PURPLE_WOOL
    - BLUE_WOOL
    - BROWN_WOOL
    - GREEN_WOOL
    - RED_WOOL
    - BLACK_WOOL
    - SNOW
    - GLASS_PANE
    - oak_leaves
    - spruce_leaves
    - birch_leaves
    - jungle_leaves
    - acacia_leaves
    - dark_oak_leaves
    - azalea_leaves
    - flowering_azalea_leaves
    - end_stone
    - grass
    - fern
    - tall_grass
    - kelp_plant
    - chorus_plant
    - oak_stairs
    - NETHERRACK
    - dirt_path
    - cobblestone_stairs
    DoorsOpenDelayInSeconds: 300
    CooldownEndInMinutes: 60
  EndermenMoveBlocks: true
  SilverfishBreakBlocks: false
  CreaturesTrampleCrops: false
  RabbitsEatCrops: true
  HardModeZombiesBreakDoors: true
  Database:
    URL: NO
    UserName: NO
    Password: NO
  UseBanCommand: true
  BanCommandPattern: ban %name% %reason%
  Advanced:
    fixNegativeClaimblockAmounts: true
    ClaimExpirationCheckRate: 60
    OfflinePlayer_cache_days: 90
  Abridged Logs:
    Days To Keep: 365
    Included Entry Types:
      Social Activity: true
      Suspicious Activity: true
      Administrative Activity: true
      Debug: true
      Muted Chat Messages: false
```

### Running without GriefPrevention

- [ ] I attempted running the server without GriefPrevention installed.
- [ ] The problem does _not_ occur when GriefPrevention is removed from the server.

### Running with only GriefPrevention

- [x] I attempted running only GriefPrevention on the server.
- [x] The issue still occurs when GriefPrevention is the only plugin running.

### Running on a fresh, clean server installation

- [x] I attempted testing for the issue on a new server.
- [x] The issue still occurs on a new server.

### Using unmodified client

- [x] I attempted testing for the issue with the vanilla client.
- [x] The issue still occurs when using the vanilla client.

### We appreciate you taking the time to fill out a bug report!

- [X] I searched for similar issues before submitting this bug report.

**Comment by Jikoo** [comment]
Sounds like time to roll a separate Siege addon. I take it you use SQL based on the `NO` for the database info? Might use that as an excuse to start fleshing out an improved DB system.

**Comment by TomLewis** [comment]
> Sounds like time to roll a separate Siege addon. I take it you use SQL based on the `NO` for the database info? Might use that as an excuse to start fleshing out an improved DB system.

Yeah I use a database ha, I have tens of thousands of claims. Not ideal for flatfile.
Please be carful with database schema changes, that requires conversion and Minecraft plugins always tend to be "lets load everything into memory in java!" and ignore that some of us have large databases that always fail with this design.
If schema has to change, always give raw SQL to manually do it too!

We forked an older version of GP with a bunch of QOL imrpovements to siege, like timers, titles, boss bar cooldowns, differnt mechanics in looting time, but as Im not a java programmer and our old developer is no longer around we wasnt able to migrate these changes we used to have! I would be more than happy to help design a new siege, I already have multiple documents outlining improvements from 9+ years of use with it!

I remember the developer saying that siege is coupled in spagetti code in GP and it would be hard to extract, we thought it would be better to seperate too.

An alernative solution would be to also create a siege plugin that supports GriefDefender so we have options going forward if GriefPrevention stays in this constant state of not being updated. I swear it was in a re-write for 3 years that never appered?

**Comment by Jikoo** [comment]
> Yeah I use a database ha, I have tens of thousands of claims. Not ideal for flatfile. Please be carful with database schema changes, that requires conversion and Minecraft plugins always tend to be "lets load everything into memory in java!" and ignore that some of us have large databases that always fail with this design. If schema has to change, always give raw SQL to manually do it too!

I have no intention of changing the schema, don't worry. The main issue is the way GP handles all of its SQL interactions. I'm not even sure that siege does store any info in the datastore, but if it does, it's a good excuse to write a better system that can then be added into GP.

GP's db implementation insists that whatever GP currently thinks is correct is the absolute truth. In addition to that, GP deletes and then writes data rather than update or modify data. This makes GP pretty incompatible with external db modification and makes it prone to data loss.
For example, say I give a user 100 bonus claim blocks via command:
A sane way to handle this would be to add 100 bonus blocks and tell the SQL db to initialize with 100 or add 100 to the existing value. This would cause a relog to refresh the user's data and correct any communication issues.
GP handles this by deleting the user's bonus blocks from storage and then in a separate call setting their bonus blocks to the current value.

> We forked an older version of GP with a bunch of QOL imrpovements to siege, like timers, titles, boss bar cooldowns, differnt mechanics in looting time, but as Im not a java programmer and our old developer is no longer around we wasnt able to migrate these changes we used to have! I would be more than happy to help design a new siege, I already have multiple documents outlining improvements from 9+ years of use with it!
> 
> I remember the developer saying that siege is coupled in spagetti code in GP and it would be hard to extract, we thought it would be better to seperate too.

Siege was hacked in on top of GP instead of exposing things to make it an addon. Iirc BigScary then decided he didn't like it on his server and it stopped being improved.

I have created the (currently empty beyond the template) [GPSiege repo](https://github.com/GPAddons/GPSiege). If you would, please open issue(s) detailing any features you'd like to see. This could range from core documentation of what siege is actually supposed to do (I have a vague concept, having shifted things around to keep it functional) to feature requests.

> An alernative solution would be to also create a siege plugin that supports GriefDefender so we have options going forward if GriefPrevention stays in this constant state of not being updated. I swear it was in a re-write for 3 years that never appered?

Not sure if that's super possible - GD is effectively a GP fork with a lot of rewrites, but siege as an addon depends heavily on rewriting effective claim permissions. It may be, would need research, but I would want to write a base plugin with that as a vague possibility in mind first.

Re: the GP rewrite - Robo realized a solo full rewrite is prohibitively difficult and it got scrapped a long while back. Personally I've got my own (mild meme) GP fork that I basically mash all the PRs I like into. When I published it I also included an [open letter to Robo](https://github.com/Easterlyn/GamerProtection#an-open-letter-to-robomwm), who reappeared the next week only to now vanish again for over a month.

**Comment by TomLewis** [comment]
Thanks for your reply!

> I have no intention of changing the schema, don't worry. The main issue is the way GP handles all of its SQL interactions. I'm not even sure that siege does store any info in the datastore, but if it does, it's a good excuse to write a better system that can then be added into GP.

It does not! and that is something I want to change with a new iteration for tracking all battles, to then be able to create levelling system, but also a balancing system so that new players can not be attacked by high level players!

> GP deletes and then writes data rather than update or modify data

This cant be right? `UPDATE` exists in SQL.

One thing that is amazing about the deisgn of the SQL that should not be tampered with is the dynamic caluclation of claim blocks left with the design choise of aquired and bonus blocks, where aquired will only ever increment and is earned by playtime, and then anything else bought/sold etc are handled by bonus blocka and then `aquired+bonus-used` gives you claim blocks left, I love this design. At one point GP was "rewritten" I think by the second developer that took over GP and they screwed with this design and broke so many things, I convinced them to revert to this way of saving.

> Siege was hacked in on top of GP instead of exposing things to make it an addon. Iirc BigScary then decided he didn't like it on his server and it stopped being improved.

Yeah it needs its own scope and settings, theres a bunch of flaws how the current version stands, but once its up and working the basics can stay the same, we have been using siege for so long now, we have a multitude of differnt flows that could work for differnt server types.

> I have created the (currently empty beyond the template) [GPSiege repo](https://github.com/GPAddons/GPSiege). If you would, please open issue(s) detailing any features you'd like to see. This could range from core documentation of what siege is actually supposed to do (I have a vague concept, having shifted things around to keep it functional) to feature requests.

Now that is good news for future proofing a siege! What would be fantastic is if we could get the current GP build just working again, so that I dont have to disable siege until this is ready to use, so we can keep players playing and then we can focus on this new version. Yesterday I had to disable siege because of the complaints from the soft blocks not working.

> Re: the GP rewrite - Robo realized a solo full rewrite is prohibitively difficult and it got scrapped a long while back. Personally I've got my own (mild meme) GP fork that I basically mash all the PRs I like into. When I published it I also included an [open letter to Robo](https://github.com/Easterlyn/GamerProtection#an-open-letter-to-robomwm), who reappeared the next week only to now vanish again for over a month.

This is why I was looking to move to [GriefDefender](https://www.spigotmc.org/resources/1-12-2-1-18-griefdefender-claim-plugin-grief-prevention-protection.68900/), maybe its time. All that is missing from it is siege and its got a solid API and its regualrly updated. Just putting it out there, maybe its time, I havent moved yet as we need siege, but on top of that hardly anyone supports it so it would need a bunch of us to move.

Either way im super happy that there is someone else that wants to revice siege for land protection in a rough way that this siege works, as I havent been able to find a relplacement solution that dousnt require moving to a differnt claim system of their own.

**Comment by Jikoo** [comment]
> > GP deletes and then writes data rather than update or modify data
> 
> This cant be right? `UPDATE` exists in SQL.
> 
> One thing that is amazing about the deisgn of the SQL that should not be tampered with is the dynamic caluclation of claim blocks left with the design choise of aquired and bonus blocks, where aquired will only ever increment and is earned by playtime, and then anything else bought/sold etc are handled by bonus blocka and then `aquired+bonus-used` gives you claim blocks left, I love this design. At one point GP was "rewritten" I think by the second developer that took over GP and they screwed with this design and broke so many things, I convinced them to revert to this way of saving.

[I really wish I was joking.](https://github.com/TechFortress/GriefPrevention/blob/0e3a2e979161327946365e71b97bbcf74f572e6b/src/main/java/me/ryanhamshire/GriefPrevention/DatabaseDataStore.java#L526-L550) The source of many SQL issues, and why I encourage people to use flatfile until SQL is rewritten. SQL _should_ be much more secure than flatfile, but the way GP does it is actually more loss-prone.

As far as I can tell the SQL is incredibly poor because they wanted to support as many SQL/NoSQL/etc. database implementations as possible without writing additional adapters. For example, an update call on old SQLite (Bukkit 1.8-1.12) is `INSERT OR REPLACE INTO x` which internally deletes and inserts, causing delete triggers to run. More modern SQLite can do `INSERT INTO x ON CONFLICT DO UPDATE` which actually properly updates values and does not call delete triggers. Both syntaxes are completely different from actual SQL syntax, which has its own variations based on implementation.

> Now that is good news for future proofing a siege! What would be fantastic is if we could get the current GP build just working again, so that I dont have to disable siege until this is ready to use, so we can keep players playing and then we can focus on this new version. Yesterday I had to disable siege because of the complaints from the soft blocks not working.

Honestly if it's busted I'll basically have to figure out what's wrong, which involves tracing through all of siege anyway. Then we'll have to wait for Robo to resurface to pull the fixes, which he'll be semi-reluctant to do because neither of us really wants siege in the base plugin. It would overall be easier to cut and paste the relevant details, particularly if none of it is in the datastore.
 
> This is why I was looking to move to [GriefDefender](https://www.spigotmc.org/resources/1-12-2-1-18-griefdefender-claim-plugin-grief-prevention-protection.68900/), maybe its time. All that is missing from it is siege and its got a solid API and its regualrly updated. Just putting it out there, maybe its time, I havent moved yet as we need siege, but on top of that hardly anyone supports it so it would need a bunch of us to move.
> 
> Either way im super happy that there is someone else that wants to revice siege for land protection in a rough way that this siege works, as I havent been able to find a relplacement solution that dousnt require moving to a differnt claim system of their own.

My personal problem with GD is the way it happened. GD is rebranded GPSponge afaik, which was a much more richly featured adoption of original GP. It went premium and claimed GP itself was dead, while the only thing that was killed was GPSponge. That left a sour taste in my mouth and I'm not willing to give them my money as a result. The [GP org on GH](https://github.com/GriefPrevention) actually is owned by the owner of GD, and I believe it used to say something to the extent of "buy GD" though it's now just empty.

From a development perspective GD is nice to work with for a premium plugin because the API is open source.

**Comment by vapecloudbear** [comment]
Hi. I was wondering if anyone else had reported issues with siege in 1.18.2?

**Comment by vapecloudbear** [comment]
To add to this bug, I have tested on purpur 1.18.2. After the siege initiator wins, buttons and lever work as expected, however, the victor cannot break blocks, and the victim/loser can place blocks but not break. I'm not requesting a rewrite, but if the base implementation could be fixed, that would be wonderful.

**Comment by TomLewis** [comment]
Bump

**Comment by Jikoo** [comment]
Siege has been removed from GP as of 8d285b5ac362fc9f63ecc1fb0a28fe4c374b8b38, so this only affects the legacy branch. As you're aware, I burned out on writing siege as an addon because it's a massive undertaking. Would be open to passing that off to another maintainer.

**Comment by RoboMWM** [comment]
I somehow skipped out on this issue when going through my notifications some months ago. 

> who reappeared the next week only to now vanish again for over a month

Yes life was crazy (still is crazy), and I'm surprised how much has changed. Though I suppose that's always been the case with me xd. Been trying to re-involve myself with development lately though. Helps that the GD guy gave me the GP org so I can have a nice clean place for all GP-core-related stuff.

> When I published it I also included an [open letter to Robo](https://github.com/Easterlyn/GamerProtection#an-open-letter-to-robomwm),

trying to find this lol, idk if I've seen it xd.

But ya, discussion thread or https://github.com/GriefPrevention/GriefPrevention/discussions/2065 for discussion of things like database stuff would be appreciated

**Comment by RoboMWM** [comment]
found the letter, yes idk where to publish my email and not get spam - and even then, I already get spammed out as-is. Wondering if I should host my own email server - but that's a discussion to be had in another place. I'll go ahead and shoot you an email tho xd.

**Comment by TomLewis** [comment]
I would absolutely love to work with anyone on getting siege made as a seperate plugin for griefprevention.

We made some progress within my team as it was a core function of piratecraft, but we never got it completely done due to life just happening for each person that took on the task!

Sadly I still don't know ant java and keep kicking myself I didn't learn it years ago when I had more time! I'm down to nothing nowadays.

Siege is a major feature for Piratecraft and since I made this ticket it's been almost a daily request asking when it's back.

There is a few core features missed such as user feedback, boss bar timers and cooldowns that need to be added to prevent bullying.

Happy to work with anyone that wants to tackle the project as it's such a beloved missed feature!
